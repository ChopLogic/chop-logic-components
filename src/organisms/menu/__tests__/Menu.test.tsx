import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it } from 'vitest';

import ChopLogicMenu from '../Menu.tsx';

describe('ChopLogicMenu Component', () => {
  const mockItems = [
    { id: '1', label: 'Item 1' },
    { id: '2', label: 'Item 2', nestedItems: [{ id: '3', label: 'Nested Item' }] },
  ];

  it('renders a menu bar with the correct role', () => {
    render(<ChopLogicMenu items={mockItems} />);
    const menuBar = screen.getByRole('menubar');
    expect(menuBar).toBeInTheDocument();
  });

  it('renders all MenuItem components for given items', () => {
    render(<ChopLogicMenu items={mockItems} />);
    const menuItems = screen.getAllByRole('menuitem');
    expect(menuItems).toHaveLength(mockItems.length);
    expect(screen.getByText('Item 1')).toBeInTheDocument();
    expect(screen.getByText('Item 2')).toBeInTheDocument();
  });

  it('should open the sub-menu onClick the item with nested items', async () => {
    render(<ChopLogicMenu items={mockItems} />);
    expect(screen.queryByText('Nested Item')).not.toBeInTheDocument();
    await userEvent.click(screen.getByText('Item 2'));
    expect(screen.getByText('Nested Item')).toBeInTheDocument();
  });

  it('should close the sub-menu on second click', async () => {
    render(<ChopLogicMenu items={mockItems} />);
    await userEvent.click(screen.getByText('Item 2'));
    expect(screen.getByText('Nested Item')).toBeInTheDocument();
    await userEvent.click(screen.getByText('Item 2'));
    expect(screen.queryByText('Nested Item')).not.toBeInTheDocument();
  });

  it('handles an empty items array gracefully', () => {
    render(<ChopLogicMenu items={[]} />);
    const menuItems = screen.queryAllByTestId('menuitem');
    expect(menuItems).toHaveLength(0);
    const menuBar = screen.getByRole('menubar');
    expect(menuBar).toBeInTheDocument();
  });
});
