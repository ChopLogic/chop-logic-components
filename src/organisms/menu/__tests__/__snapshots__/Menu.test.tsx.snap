// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu Component > renders a menu bar with the correct role 1`] = `
<DocumentFragment>
  <menu
    class="_menuBar_b8ce27 _menuBar__vertical_b8ce27"
  >
    <li
      class="_menuLeaf_f4b00b"
      role="menuitem"
      tabindex="0"
    >
      <span>
        Item 1
      </span>
    </li>
    <li
      aria-expanded="false"
      aria-haspopup="true"
      class="_subMenu_0275f8"
      role="menuitem"
      tabindex="0"
    >
      <span
        class="_text_0275f8"
      >
        <span>
          Item 2
        </span>
        <svg
          class="_icon_3b2ece"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M21.886 5.536A1.002 1.002 0 0021 5H3a1.002 1.002 0 00-.822 1.569l9 13a.998.998 0 001.644 0l9-13a.998.998 0 00.064-1.033zM12 17.243L4.908 7h14.184L12 17.243z"
          />
        </svg>
      </span>
    </li>
  </menu>
</DocumentFragment>
`;
